rust/template/distributed_datalog/src/instantiate.rs

```diff
@@ -405,13 +301,218 @@ where
+            Sink::File(path) => {
+                let file = File::create(path)
+                    .map_err(|e| format!("failed to create file {}:, {}", path.display(), e))?;
+                let file_sink = Arc::new(Mutex::new(FileSink::<P::Convert>::new(file)));
+
+                // Subscribe the accumulator to this sink.
+                let subscription = accumulator
+                    .lock()
+                    .unwrap()
+                    .subscribe(Box::new(file_sink.clone()))
+                    .map_err(|_| "failed to subscribe file sink to accumulator".to_string())?;
+
+                // Add sink to sink map for this accumulator.
+                let _ = sink_map.insert(
+                    sink.clone(),
+                    (SinkRealization::File(file_sink), subscription),
+                );
+            }
+            Sink::TcpSender(addr) => {
+                match addr {
+                    Addr::Ip(address) => {
+                        let tcp_sender = TcpSender::new(*address)
+                            .map_err(|e| format!("failed to create TcpSender socket: {}", e))?;
+                        let sink = Arc::new(Mutex::new(tcp_sender));
+
+                        // Subscribe the accumulator to this sink.
+                        let subscription = accumulator
+                            .lock()
+                            .unwrap()
+                            .subscribe(Box::new(sink.clone()))
+                            .map_err(|_| "failed to subscribe file sink to accumulator")?;
+
+                        // Add sink to sink map for this accumulator.
+                        let _ = sink_map.insert(
+                            Sink::TcpSender(*addr),
+                            (SinkRealization::Node(sink), subscription),
+                        );
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+
+    /// Remove the sink accumulator from the realization.
+    /// Remove accumulator's entry (keyed by rel_ids) from _sinks.
+    /// Disconnect accumulator from the server:
+    /// - Unsubscribe the stream from the accumulator.
+    /// - Remove the stream from the server.
+    /// Clear the accumulator.
+    pub fn remove_sink_accumulator(
+        &mut self,
+        rel_ids: BTreeSet<RelId>,
+        server: &mut DDlogServer<P>,
+    ) -> Result<(), String> {
+        if let Some(entry) = self._sinks.remove(&rel_ids) {
+            let (accumulator, mut stream, sink_map) = entry;
+            if sink_map.is_empty() {
+                // Disconnect accumulator from server.
+                // First, unsubscribe the stream from the accumulator.
+                let _ = stream.unsubscribe(&());
+                // Next, remove the stream from the server.
+                server.remove_stream(stream);
+
+                let mut accum = accumulator.lock().unwrap();
+                // Clear the accumulator.
+                if accum.clear().is_ok() {
```

Comment:
`accum` is the last reference to this accumulator, right? So shouldn't it clear itself internal state as it goes out of scope?